packages/react-query/src/shared/hooks/createHooksInternal.tsx

```diff
@@ -273,24 +281,30 @@ export function createRootHooks<
     const shouldAbortOnUnmount =
       opts?.trpc?.abortOnUnmount ?? config?.abortOnUnmount ?? abortOnUnmount;
 
-    const hook = __useQuery({
-      ...ssrOpts,
-      queryKey: queryKey as any,
-      queryFn: (queryFunctionContext) => {
-        const actualOpts = {
-          ...ssrOpts,
-          trpc: {
-            ...ssrOpts?.trpc,
-            ...(shouldAbortOnUnmount
-              ? { signal: queryFunctionContext.signal }
-              : {}),
-          },
-        };
+    const hook = __useQuery(
+      {
+        ...ssrOpts,
+        queryKey: queryKey as any,
+        queryFn: async (queryFunctionContext) => {
+          const actualOpts = {
+            ...ssrOpts,
+            trpc: {
+              ...ssrOpts?.trpc,
+              ...(shouldAbortOnUnmount
+                ? { signal: queryFunctionContext.signal }
+                : {}),
+            },
+          };
 
-        return client.query(...getClientArgs(queryKey, actualOpts));
+          const res = await client.query(
+            ...getClientArgs(queryKey, actualOpts),
+          );
+          opts?.trpc?.__internal_afterQueryFn?.();
```

Comment:
this is instead of the `useQuery.onSuccess` used for testing.

open for suggestions on how to make this redundant